File: server/server_main/app/services/command_service.py
from ..extensions import db
from ..models.device_command import DeviceCommand


def queue_command(kiosk_id: int, command_type: str, payload: dict = None) -> DeviceCommand:
    """Create a PENDING command for the ESP32 to pick up on next poll."""
    cmd = DeviceCommand(kiosk_id=kiosk_id, command_type=command_type)
    if payload:
        cmd.set_payload(payload)
    db.session.add(cmd)
    db.session.flush()  # get id before commit
    return cmd


def get_pending_commands(kiosk_id: int) -> list[DeviceCommand]:
    """Return all PENDING commands for a kiosk, oldest first."""
    return (
        DeviceCommand.query.filter_by(kiosk_id=kiosk_id, status="PENDING")
        .order_by(DeviceCommand.created_at.asc())
        .all()
    )


def ack_command(command_id: int, kiosk_id: int) -> DeviceCommand | None:
    """Mark a command as ACKED. Returns None if not found or wrong kiosk."""
    cmd = DeviceCommand.query.filter_by(
        id=command_id, kiosk_id=kiosk_id, status="PENDING"
    ).first()
    if cmd:
        cmd.ack()
        db.session.flush()
    return cmd

File: server/server_main/src/services/commandService.ts
import prisma from "../prisma";

const COMMAND_TTL_MS = 5 * 60 * 1000; // 5 minutes — commands older than this are expired

export async function queueCommand(
  kioskId: number,
  commandType: string,
  payload?: object,
) {
  return prisma.deviceCommand.create({
    data: {
      kioskId,
      commandType,
      payload: payload ? JSON.stringify(payload) : null,
    },
  });
}

export async function getPendingCommands(kioskId: number) {
  // Auto-expire stale PENDING commands so the ESP is never flooded with old test commands
  const expireBefore = new Date(Date.now() - COMMAND_TTL_MS);
  await prisma.deviceCommand.updateMany({
    where: { kioskId, status: "PENDING", createdAt: { lt: expireBefore } },
    data: { status: "EXPIRED" },
  });

  return prisma.deviceCommand.findMany({
    where: { kioskId, status: "PENDING" },
    orderBy: { createdAt: "asc" },
    take: 5, // max 5 commands per poll — prevents serial ack backlog on the ESP
  });
}

export async function ackCommand(commandId: number, kioskId: number) {
  return prisma.deviceCommand.updateMany({
    where: { id: commandId, kioskId, status: "PENDING" },
    data: { status: "ACKED", ackedAt: new Date() },
  });
}

export async function cancelKioskCommands(kioskId: number) {
  return prisma.deviceCommand.updateMany({
    where: { kioskId, status: "PENDING" },
    data: { status: "EXPIRED" },
  });
}
